firefox_puppeteer/tests/test_toolbars.py

```diff
@@ -0,0 +1,135 @@
+        event_types = ["shortcut", "shortcut2", "button"]
+        for event in event_types:
+            # TODO: Until we have waitForPageLoad, this only tests API
+            # compatibility.
+            self.browser.navbar.locationbar.reload_url(event, force=True)
+            self.browser.navbar.locationbar.reload_url(event, force=False)
+
+    def test_focus_and_clear(self):
+        locationbar = self.browser.navbar.locationbar
+        locationbar.urlbar.send_keys("zyx")
+        locationbar.clear()
+        self.assertEqual(locationbar.value, '')
+        locationbar.urlbar.send_keys("zyx")
+        self.assertEqual(locationbar.value, 'zyx')
+        locationbar.clear()
+        self.assertEqual(locationbar.value, '')
+
+    def test_load_url(self):
+        data_uri = 'data:text/html,<title>Title</title>'
+        locationbar = self.browser.navbar.locationbar
+        locationbar.load_url(data_uri)
+
+        with self.marionette.using_context('content'):
+            self.wait_for_condition(lambda mn: mn.get_url() == data_uri)
+
+    def test_urlbar_input(self):
+        urlbar_input = self.browser.navbar.locationbar.urlbar_input
+        self.assertEqual('input', urlbar_input.get_attribute('localName'))
+        self.assertIn('urlbar-input', urlbar_input.get_attribute('className'))
+
+    def test_contextmenu(self):
+        contextmenu = self.browser.navbar.locationbar.contextmenu
+        self.assertEqual('menupopup', contextmenu.get_attribute('localName'))
+
+    def test_contextment_entry(self):
+        contextmenu_entry = self.browser.navbar.locationbar.get_contextmenu_entry('paste')
+        self.assertEqual('cmd_paste', contextmenu_entry.get_attribute('cmd'))
+
+    def test_reloadbutton(self):
+        reload_button = self.browser.navbar.locationbar.reload_button
+        self.assertEqual('Browser:ReloadOrDuplicate', reload_button.get_attribute('command'))
+
+    def test_urlbar(self):
+        urlbar = self.browser.navbar.locationbar.urlbar
+        self.assertEqual('textbox', urlbar.get_attribute('localName'))
+
+    def test_favicon(self):
+        favicon = self.browser.navbar.locationbar.favicon
+        self.assertEqual('image', favicon.get_attribute('localName'))
+
+    def test_history_drop_marker(self):
+        drop_marker = self.browser.navbar.locationbar.history_drop_marker
+        self.assertEqual('dropmarker', drop_marker.get_attribute('localName'))
+
+    def test_stopbutton(self):
+        stop_button = self.browser.navbar.locationbar.stop_button
+        self.assertEqual('toolbarbutton', stop_button.get_attribute('localName'))
+
+
+class TestAutoCompleteResults(FirefoxTestCase):
+    def setUp(self):
+        FirefoxTestCase.setUp(self)
+        self.browser.navbar.locationbar.clear()
+
+    def tearDown(self):
+        try:
+            autocompleteresults = self.browser.navbar.locationbar.autocomplete_results
+            autocompleteresults.close(force=True)
+        except NoSuchElementException:
```

Comment:
Where does this fail exactly? in close()? I think we should make this failure proof so it only operates on the elemnet if it really exists, especially in the case of an exception.